matrix/buffer.py

```diff
@@ -703,24 +723,44 @@ def join(self, user, date, message=True, extra_tags=None):
             self.print_date_tags(msg, date, tags)
             self.add_smart_filtered_nick(user.nick)
 
-    def invite(self, nick, date, extra_tags=None):
+    def change_nick(self, user_id, old_nick, new_nick, date, message=True, extra_tags=None):
+        # type: (str, str, int, Optional[bool], Optional[List[str]]) -> None
+        user = self._get_user(user_id)
+        self._remove_user_from_nicklist(user)
+        user.nick = new_nick
+        self._add_user_to_nicklist(user)
+
+        if message:
+            tags = self._message_tags(user, "change")
+            msg = self._nick_change_message(user, old_nick, new_nick)
+
+            # TODO add a option to disable smart filters
+            tags.append(SCRIPT_NAME + "_smart_filter")
+
+            self.print_date_tags(msg, date, tags)
+            self.add_smart_filtered_nick(user.nick)
```

Comment:
You need to remove the old nick using remove_smart_filtered_nick(). The smart filter algorithm should then pick the old nick up from the tags just like it does in the irc plugin. 

You don't need to make the smart filter algorithm work with nick changes for now. Just make sure to add the tags and remove the old nick here.